### PR TITLE
Add animations and trigger on hover

### PR DESCRIPTION
**Pull request by Klynger**
#### What does this PR do? \*

- Add support to trigger on hover.
- Add support to animations.

#### How to test it? \*

[workspace with animation](https://popoveranimations--storecomponents.myvtex.com/)
[workspace without animation](https://popoverlayout--storecomponents.myvtex.com/)

### CODE REVIEW

File: react/components/Animations/Fade.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import React from 'react'
```
Comment (Klynger) [inline react/components/Animations/Fade.tsx:1]: This is the same implementation of the [Fade component of modal-layout](https://github.com/vtex-apps/modal-layout/blob/master/react/components/Animations/Fade.tsx). I'm waiting to see another place that I have to implement this animation and implement another type of animation to create a module of animations

File: react/Popover.tsx [hunk 2]
```diff
@@ -1,62 +1,113 @@
...
+        dispatch({ type: 'CLOSE_POPOVER' })
+      }
+    },
+    [dispatch, triggerMode]
+  )
+
+  const handleClick = useCallback(
+    (e: React.MouseEvent) => {
+      if (triggerMode === 'click') {
+        e.stopPropagation()
```
Comment (klzns) [inline react/Popover.tsx:77]: Why is that needed?

File: react/Trigger.tsx [hunk 3]
```diff
@@ -35,8 +36,17 @@ function Trigger(props: Props) {
     }
   }, [dispatch, containerRef])
 
+  useEffect(() => {
+    if (dispatch) {
```
Comment (klzns) [inline react/Trigger.tsx]: Why is that needed?

File: react/Trigger.tsx [hunk 4]
```diff
@@ -47,11 +57,35 @@ function Trigger(props: Props) {
   }
 
   const handleClick = (e: React.MouseEvent<TriggerElement>) => {
+    if (trigger !== 'click') {
+      return
+    }
+
     e.stopPropagation()
```
Comment (klzns) [inline react/Trigger.tsx:66]: Why is that needed?

File: react/components/Popper.tsx [hunk 5]
```diff
@@ -174,7 +174,7 @@ const Popper = forwardRef(function Popper(props: Props, ref) {
       <div
         role={role}
         ref={handleRef}
-        style={{ position: 'fixed', left: 0, top: 0 }}
+        style={{ position: 'fixed', left: 0, top: 0, ...rest.style }}
```
Comment (klzns) [inline react/components/Popper.tsx]: Remove style from object, or this prop will be overwritten

**Review by klzns — CHANGES_REQUESTED** [review]

File: react/Popover.tsx [hunk 6]
```diff
@@ -1,62 +1,113 @@
...
+    case 'fade':
+      return Fade
+    default:
+      return Fade
+  }
+}
+
+function useTransition(option: TransitionComponentType) {
+  return useMemo(() => getTransitionComponent(option), [option])
 }
```
Comment (victorhmp) [inline react/Popover.tsx]: These could be the same function. Maybe move `getTransitionComponent` into `useTransition`. If you want you could even create a `hooks` folder in this project and move `useTransition` there so that this logic is completely abstracted in the context of `Popover.tsx`.

File: react/Trigger.tsx [hunk 7]
```diff
@@ -47,11 +57,35 @@ function Trigger(props: Props) {
...
 
     if (dispatch) {
-      dispatch({ type: 'OPEN_POPOVER' })
+      if (open) {
+        dispatch({ type: 'CLOSE_POPOVER' })
+      } else {
+        dispatch({ type: 'OPEN_POPOVER' })
+      }
+    }
+  }
```
Comment (victorhmp) [inline react/Trigger.tsx]: I would check for`!dispatch` or `dispatch == null` and return early in that case.
Removes at least one nesting level and the "happy path" is not wrapped in a conditional

File: react/Trigger.tsx [hunk 8]
```diff
@@ -47,11 +57,35 @@ function Trigger(props: Props) {
...
+
+  const handleMouseEnter = () => {
+    if (trigger === 'hover') {
+      if (dispatch) {
+        dispatch({ type: 'OPEN_POPOVER' })
+      }
+    }
```
Comment (victorhmp) [inline react/Trigger.tsx]: This condition could be `if (!dispatch || trigger !== 'hover') return` (not in the same line hehe)

File: react/Popover.tsx [hunk 9]
```diff
@@ -1,62 +1,113 @@
...
+
+  const handleKeydown = useCallback(
+    (e: React.KeyboardEvent<HTMLDivElement>) => {
+      if (e.key !== 'Escape' || triggerMode !== 'click') {
+        return
+      }
+
+      if (dispatch) {
+        dispatch({ type: 'CLOSE_POPOVER' })
+      }
```
Comment (victorhmp) [inline react/Popover.tsx]: This check for `dispatch` being truthy could be moved up into the conditional above since you would do nothing in this case.

File: react/components/PopoverContext.tsx [hunk 10]
```diff
@@ -48,6 +59,11 @@ function popoverContextReducer(state: State = DEFAULT_STATE, action: Action) {
         ...state,
         containerRef: action.payload?.containerRef,
       }
+    case 'SET_TRIGGER_MODE':
+      return {
+        ...state,
+        triggerMode: action.payload?.triggerMode,
```
Comment (victorhmp) [inline react/components/PopoverContext.tsx]: This could cause you to set `triggerMode` to `undefined`, right? That might not be what you want.

**Review by victorhmp — CHANGES_REQUESTED** [review]

File: react/Popover.tsx (hunk 2 above)
Comment (Klynger) [reply react/Popover.tsx:77]: To prevent the event from bubbling up to the trigger, because the trigger is a toggle (is expected that if you click in the trigger with the popover opened it should close).

File: react/Popover.tsx [hunk 11]
```diff
@@ -1,62 +1,113 @@
...
-  const { children, placement, role } = props
-  const { open, containerRef } = usePopoverState()
+  const { children, placement, role, transitionComponent = 'fade' } = props
+  const { open, containerRef, triggerMode } = usePopoverState()
   const dispatch = usePopoverDispatch()
   const handles = useCssHandles(CSS_HANDLES)
+  const TransitionComponent = useTransition(transitionComponent)
 
-  const classes = classnames(
+  const paperClasses = classnames(
```
Comment (iaronaraujo) [inline react/Popover.tsx]: What's `paperClasses`? Is there a reason for this name?

File: react/Trigger.tsx (hunk 4 above)
Comment (Klynger) [reply react/Trigger.tsx:66]: Because I'm committing something to be done with this event already, so it seems to me that if I let the event go further some parent element might get the event and do something too (like opening a modal, going to some link, etc)

File: react/Trigger.tsx (hunk 3 above)
Comment (iaronaraujo) [reply react/Trigger.tsx]: Add a comment here

File: react/Popover.tsx (hunk 2 above)
Comment (Klynger) [reply react/Popover.tsx:77]: I think I should add a comment to that

File: react/Trigger.tsx (hunk 3 above)
Comment (Klynger) [reply react/Trigger.tsx]: `useContext` returns `T | undefined`

File: react/Popover.tsx (hunk 11 above)
Comment (Klynger) [reply react/Popover.tsx]: This is the element where you put the background of the popover, which looks like a paper

File: react/Trigger.tsx (hunk 3 above)
Comment (Klynger) [reply react/Trigger.tsx]: I changed the type to always return a function

File: react/Popover.tsx [hunk 12]
```diff
@@ -1,62 +1,110 @@
...
```
Comment (iaronaraujo) [inline react/Popover.tsx]: rename this

**Review by iaronaraujo — CHANGES_REQUESTED** [review]

File: react/Trigger.tsx [hunk 13]
```diff
@@ -1,60 +1,90 @@
...
-    if (dispatch) {
+    if (open) {
+      dispatch({ type: 'CLOSE_POPOVER' })
+    } else {
+      dispatch({ type: 'OPEN_POPOVER' })
+    }
+  }
+
+  const handleMouseEnter = () => {
+    if (trigger === 'hover') {
```
Comment (klzns) [inline react/Trigger.tsx]: I think you could just not add this function handler on onMouseEnter if it's not hover

File: react/components/TrapFocus.tsx [hunk 14]
```diff
@@ -1,19 +1,33 @@
...
   useEffect(() => {
-    if (open && childRef.current) {
-      childRef.current.focus()
+    if (open) {
+      // Popperjs probably always start the element at the top of the screen.
+      // So if I try to focus after it position the element in the right place
+      // the scroll will go to the top of the page. That's why I have to use
+      // the requestAnimationFrame
+      requestAnimationFrame(() => {
+        // And we need two because of IE11
```
Comment (klzns) [inline react/components/TrapFocus.tsx:22]: 😩

**Review by klzns — CHANGES_REQUESTED** [review]

File: react/Trigger.tsx [hunk 15]
```diff
@@ -1,60 +1,92 @@
...
 const CSS_HANDLES = ['trigger']
 
-// TODO hover trigger
-type TriggerMode = 'click' | 'hover'
+export type TriggerMode = 'click' | 'hover'
 
 interface Props {
   children: React.ReactNode
-  trigger?: TriggerMode
+  trigger?: MaybeResponsiveInput<TriggerMode>
```
Comment (victorhmp) [inline react/Trigger.tsx:21]: I see why using responsive-values could be useful, but I also think that you should not allow for `'hover'` on mobile

**Review by klzns — APPROVED** [review]

**Review by victorhmp — CHANGES_REQUESTED** [review]

**Review by victorhmp — APPROVED** [review]

**Review by kaisermann — APPROVED** [review]

**Review by iaronaraujo — APPROVED** [review]
